ot3_state_manager/CMakeLists.txt

```diff
@@ -0,0 +1,77 @@
+	OUTPUT ${GENERATED_MESSAGE_HEADER}
+	COMMENT "Generating header file and storing to ${GENERATED_MESSAGE_HEADER}"
+	VERBATIM
+	DEPENDS ${PYSRC_DIR}/messages.py
+	COMMAND ${Poetry_EXECUTABLE} run python3 generate_state_manager_headers.py ${GENERATED_MESSAGE_HEADER}
+	WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+	)
+add_custom_target(state-manager-headers DEPENDS ${GENERATED_MESSAGE_HEADER})
+# By depending on this target, you'll be able to include the generated file
+# with no path qualifiers (i.e. #include "ot3_state_manager_messages.hpp")
+add_library(ot3_state_manager INTERFACE)
+add_dependencies(ot3_state_manager state-manager-headers)
+target_include_directories(ot3_state_manager INTERFACE ${CMAKE_CURRENT_BINARY_DIR})
+
+add_custom_target(
+	ot3-state-manager-setup
+	COMMAND ${Poetry_EXECUTABLE} env use 3.8
+	COMMAND ${Poetry_EXECUTABLE} update
+	COMMAND ${Poetry_EXECUTABLE} install
+	WORKING_DIRECTORY ../../ot3_state_manager
+)
+
+add_custom_target(
+	ot3-state-manager-teardown
+	COMMAND ${Poetry_EXECUTABLE} env info -p | xargs basename | xargs ${Poetry_EXECUTABLE} env remove
+	WORKING_DIRECTORY ../../ot3_state_manager
+)
+
+add_custom_target(
+	ot3-state-manager-build
+	COMMAND ${Poetry_EXECUTABLE} build
+	WORKING_DIRECTORY ../../ot3_state_manager
+)
+
+add_custom_target(
+	ot3-state-manager-clean
+	COMMAND rm -rf build dist .coverage coverage.xml '*.egg-info' '__pycache__' '*.pyc' '.mypy_cache'
+	WORKING_DIRECTORY ../../ot3_state_manager
+)
+
+add_custom_target(
+	ot3-state-manager-test
+	COMMAND ${Poetry_EXECUTABLE} run py.test tests --cov=ot3_state_manager --cov-report term-missing:skip-covered --cov-report xml:coverage.xml
+	WORKING_DIRECTORY ../../ot3_state_manager
+)
+
+add_custom_target(
+	ot3-state-manager-lint
+	COMMENT "Running linting on ot3_state_manager project"
+    COMMAND ${Poetry_EXECUTABLE} run python -m mypy ot3_state_manager tests/ generate_state_manager_headers.py
+	COMMAND ${Poetry_EXECUTABLE} run python -m isort --check ot3_state_manager tests generate_state_manager_headers.py
+	COMMAND ${Poetry_EXECUTABLE} run python -m black --check --line-length 88 ot3_state_manager tests generate_state_manager_headers.py
+	COMMAND ${Poetry_EXECUTABLE} run python -m flake8 ot3_state_manager tests generate_state_manager_headers.py
+	COMMAND ${Poetry_EXECUTABLE} run python -m mdformat --check README.md
+	WORKING_DIRECTORY ../../ot3_state_manager
+)
+
+
+add_custom_target(
+	ot3-state-manager-format
+	# Have to wrap this isort command in a bash -c for some reason because otherwise it fails on the -m for no reason
+	# other than cmake is dumb sometimes.
+	bash -c "${Poetry_EXECUTABLE} run python -m isort ot3_state_manager tests generate_state_manager_headers.py"
+	${Poetry_EXECUTABLE} run python -m black ot3_state_manager tests generate_state_manager_headers.py
+	${Poetry_EXECUTABLE} run python -m mdformat README.md
+	WORKING_DIRECTORY ../../ot3_state_manager
+)
+
+add_dependencies(ot3-state-manager-test ot3-state-manager-setup)
```

Comment:
these dependencies mean that the setup task will be run whenever you run any of the other ones - it never gets saved as "already did that". it's like make in that way. as with the makefiles in the monorepo, we shouldn't explicitly have a `setup` phony target (which is what `add_custom_target` creates) as dependency to other dtargetrs.

Reply:
I am not understanding how the `setup` target is phony since it is defined above in the file.

Reply:
c803840c2aa3a2a59cf199437f4b969c247896ae

Reply:
So how do you square this circle, and make sure that
- Before you run anything in poetry, `poetry install` has been run
- But you don't run `poetry install` unnecessarily?
Well, there's two ways. One way is the build-time way, and the other way is the configure-time way.

The build-time way is more complex, but means you could run `poetry install` without doing a full project reconfigure. What you'd do is actually use _two_ commands: [`add_custom_command`](https://cmake.org/cmake/help/latest/command/add_custom_command.html) and `add_custom_target`.

First, you'd use `add_custom_command` to tell cmake about the project virtualenv that poetry creates:
```
# this must be the name of a _file_; pick something that will always be in the virtualenv like its activation script.
# we know where the virtualenv will be because we're using the local config to set it. it's probably a good idea
# to put this file path in a variable since we'll need it later
set(virtualenv_file path/to/some/file/in/virtualenv/like/bin/activate})

add_custom_command(
    # by using a custom command with an explicit output, cmake can look at the output file to decide if the command needs
    # to run
    OUTPUT ${virtualenv_file}
    # by telling cmake this command depends on pyproject.toml and poetry.lock, we can get cmake to rerun it if
    # they change
    DEPENDS pyproject.toml poetry.lock
    # the same commands you're using in add_custom_target go here
    COMMAND poetry install
)
```
The file that you use to signify "this virtualenv is built" is now something you put as a dependency in future `add_custom_target`s (this can kind of mess up parallel runs but that's fine for this use case), so
```
add_custom_target(
   ot3-state-manager-test
   COMMANDS poetry py.test
   DEPENDS ${virtualenv_file}
)
```

The configure-time way is actually much easier and fits better into cmake's view of the world. In cmake's view of the world, "configuration" is what you do to make your project ready to run _just the commands that actually make code_. In that view, `poetry install` is properly part of configuration, because `poetry run py.test` and `poetry run pylint` and `poetry build` are the commands that actually build code. By making the install happen during configuration time, you can be completely certain that it has been run by the time you run `cmake --build --target ot3-state-manager-test`. What you'd do is... call `poetry install` using [execute_process](https://cmake.org/cmake/help/latest/command/execute_process.html). That's it! This will run when you run `cmake` (note: _not_ `cmake --build`). Rerunning `cmake` will rerun it. That's all. The downside is that to re-install, you have to run `cmake` again, and this doesn't have dependency detection so it will run whenever you run `cmake`. I'd try the build time way first.